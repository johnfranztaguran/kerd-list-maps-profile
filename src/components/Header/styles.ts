import { StyleSheet, ViewStyle } from 'react-native'

type Styles = {
  mainContainer: ViewStyle
  navigationContainer: ViewStyle
  mainHeaderContent: ViewStyle
  headerLeft: ViewStyle
  headerRight: ViewStyle
  backButtonWrapper: ViewStyle
  backButton: ViewStyle
}

const styles: Styles = StyleSheet.create<Styles>({
  mainContainer: {},
  navigationContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    alignSelf: 'center',
    justifyContent: 'space-evenly',
    height: 50,
    marginLeft: 16,
    marginRight: 16,
    marginBottom: 8,
    marginTop: 8
  },
  mainHeaderContent: {
    flex: 1,
    marginHorizontal: 50,
    flexDirection: 'column',
    justifyContent: 'center'
  },
  headerLeft: {
    position: 'absolute',
    left: 0
  },
  headerRight: {
    position: 'absolute',
    right: 0
  },
  backButtonWrapper: {
    marginTop: 24,
    marginRight: 24,
    marginBottom: 24
  },
  backButton: {
    alignItems: 'flex-start',
    margin: 0
  }
})

export default styles
