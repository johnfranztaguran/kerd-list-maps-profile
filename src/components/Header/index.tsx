import React, { memo } from 'react'
import {
  View,
  StatusBar as RNStatusBar,
  Platform,
  ViewStyle,
  StyleProp,
  TouchableOpacity,
} from 'react-native'
import SafeAreaView from 'react-native-safe-area-view'
import Icon from 'react-native-vector-icons/SimpleLineIcons'
import { testIdProps } from 'utils/testAutomation'
import { strings, themes, colors } from 'constant'
import CustomStatusBar from '../StatusBar'
import Text from '../Text'
import styles from './styles'

type HeaderProps = {
  backgroundColor?: string
  isBack?: boolean
  title: string
  subtitle?: string
  statusBarStyle?: 'dark-content' | 'default' | 'light-content'
  headerStyle?: ViewStyle | any
  leftComponent?: () => JSX.Element
  rightComponent?: () => JSX.Element
  onPressBack?: () => void
}

const Header = memo(
  (props: HeaderProps): JSX.Element => {
    const {
      backgroundColor = themes.colors.primary,
      isBack = true,
      title,
      subtitle = '',
      statusBarStyle = 'light-content',
      leftComponent,
      rightComponent,
      headerStyle,
      onPressBack
    } = props

    return (
      <SafeAreaView
        {...testIdProps(
          `${strings.testIdText.prefixView}-header-main-container`
        )}
        style={[
          styles.mainContainer,
          {
            backgroundColor
          },
          ...(headerStyle ? [headerStyle] : [])
        ]}>
        <StatusBar
          backgroundColor={backgroundColor}
          statusBarStyle={statusBarStyle}
          headerStyle={headerStyle}
        />
        <View
          {...testIdProps(
            `${strings.testIdText.prefixView}-navigation-container`
          )}
          style={[styles.navigationContainer]}>
          <View
            {...testIdProps(
              `${strings.testIdText.prefixView}-header-left-container`
            )}
            style={styles.headerLeft}>
            {isBack && (
              <TouchableOpacity onPress={onPressBack}>
                <View style={styles.backButtonWrapper}>
                  <Icon name="arrow-left" size={20} color={colors.white} />
                </View>
              </TouchableOpacity>
            )}

            {leftComponent && leftComponent()}
          </View>
          <View
            {...testIdProps(
              `${strings.testIdText.prefixView}-header-title-container`
            )}
            style={styles.mainHeaderContent}>
            {subtitle !== '' ? (
              <Text
                h5
                white
                text={title}
                ellipsizeMode="tail"
                numberOfLines={1}
                medium
              />
            ) : (
              <Text
                h4
                white
                text={title}
                ellipsizeMode="tail"
                numberOfLines={1}
                medium
              />
            )}

            {subtitle !== '' && (
              <Text
                subtitle
                white
                text={subtitle}
                h4
                ellipsizeMode="tail"
                numberOfLines={1}
                light
              />
            )}
          </View>
          <View
            {...testIdProps(
              `${strings.testIdText.prefixView}-header-right-container`
            )}
            style={styles.headerRight}>
            {rightComponent && rightComponent()}
          </View>
        </View>
      </SafeAreaView>
    )
  }
)

type StatusBarProps = {
  backgroundColor: string
  statusBarStyle: 'dark-content' | 'default' | 'light-content'
  headerStyle?: StyleProp<ViewStyle>
}

const StatusBar = memo(
  (props: StatusBarProps): JSX.Element => {
    const { backgroundColor, statusBarStyle, headerStyle } = props
    const height = Platform.OS === 'ios' ? 0 : RNStatusBar.currentHeight

    return (
      <View
        {...testIdProps(`${strings.testIdText.prefixView}-statusbar-container`)}
        style={[
          {
            backgroundColor,
            height
          },
          ...(headerStyle ? [headerStyle] : [])
        ]}>
        <CustomStatusBar
          translucent
          backgroundColor={backgroundColor}
          statusBarStyle={statusBarStyle}
        />
      </View>
    )
  }
)

export default Header
