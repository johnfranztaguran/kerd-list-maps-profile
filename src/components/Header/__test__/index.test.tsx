import React from 'react'
import { Text as MockText } from 'react-native'
import { render } from '@testing-library/react-native'
import Header from '../index'

describe('Component: Header', () => {
  const onPress = jest.fn()
  it('Should render a <Header />', () => {
    const { baseElement } = render(
      <Header title="Test Title" onPressBack={onPress} />
    )
    expect(baseElement).toMatchSnapshot()
  })

  it('Should render a <Header /> with background color and right component', () => {
    const rightComponent = () => <MockText>Edit</MockText>
    const { baseElement } = render(
      <Header
        backgroundColor="red"
        title="Test Title"
        rightComponent={rightComponent}
        onPressBack={onPress}
      />
    )
    expect(baseElement).toMatchSnapshot()
  })
})
