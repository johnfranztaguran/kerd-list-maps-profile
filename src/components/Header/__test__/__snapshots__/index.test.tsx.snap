// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Header Should render a <Header /> 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityLabel="view--header-main-container"
      accessible={true}
      pointerEvents="box-none"
      style={
        Object {
          "backgroundColor": "#38647F",
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 20,
        }
      }
      testID="view--header-main-container"
    >
      <View
        accessibilityLabel="view--statusbar-container"
        accessible={true}
        style={
          Array [
            Object {
              "backgroundColor": "#38647F",
              "height": 0,
            },
          ]
        }
        testID="view--statusbar-container"
      />
      <View
        accessibilityLabel="view--navigation-container"
        accessible={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "space-evenly",
              "marginBottom": 8,
              "marginLeft": 16,
              "marginRight": 16,
              "marginTop": 8,
            },
          ]
        }
        testID="view--navigation-container"
      >
        <View
          accessibilityLabel="view--header-left-container"
          accessible={true}
          style={
            Object {
              "left": 0,
              "position": "absolute",
            }
          }
          testID="view--header-left-container"
        >
          <TouchableOpacity
            activeOpacity={0.2}
          >
            <View
              style={
                Object {
                  "marginBottom": 24,
                  "marginRight": 24,
                  "marginTop": 24,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#FFFFFF",
                      "fontSize": 20,
                    },
                    undefined,
                    Object {
                      "fontFamily": "simple-line-icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
          </TouchableOpacity>
        </View>
        <View
          accessibilityLabel="view--header-title-container"
          accessible={true}
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
              "marginHorizontal": 50,
            }
          }
          testID="view--header-title-container"
        >
          <Text
            accessibilityLabel="txt-test-title"
            accessible={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "#000000",
                    "fontSize": 12,
                  },
                  undefined,
                  undefined,
                  undefined,
                  Object {
                    "fontFamily": "OpenSans",
                    "fontSize": 18,
                    "lineHeight": 21,
                  },
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  false,
                  undefined,
                  undefined,
                  false,
                  false,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Object {
                    "fontWeight": "500",
                  },
                  undefined,
                  undefined,
                  undefined,
                  false,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Object {
                    "color": "#FFFFFF",
                  },
                  undefined,
                ],
              ]
            }
            testID="txt-test-title"
          >
            Test Title
          </Text>
        </View>
        <View
          accessibilityLabel="view--header-right-container"
          accessible={true}
          style={
            Object {
              "position": "absolute",
              "right": 0,
            }
          }
          testID="view--header-right-container"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Component: Header Should render a <Header /> with background color and right component 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityLabel="view--header-main-container"
      accessible={true}
      pointerEvents="box-none"
      style={
        Object {
          "backgroundColor": "red",
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 20,
        }
      }
      testID="view--header-main-container"
    >
      <View
        accessibilityLabel="view--statusbar-container"
        accessible={true}
        style={
          Array [
            Object {
              "backgroundColor": "red",
              "height": 0,
            },
          ]
        }
        testID="view--statusbar-container"
      />
      <View
        accessibilityLabel="view--navigation-container"
        accessible={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "space-evenly",
              "marginBottom": 8,
              "marginLeft": 16,
              "marginRight": 16,
              "marginTop": 8,
            },
          ]
        }
        testID="view--navigation-container"
      >
        <View
          accessibilityLabel="view--header-left-container"
          accessible={true}
          style={
            Object {
              "left": 0,
              "position": "absolute",
            }
          }
          testID="view--header-left-container"
        >
          <TouchableOpacity
            activeOpacity={0.2}
          >
            <View
              style={
                Object {
                  "marginBottom": 24,
                  "marginRight": 24,
                  "marginTop": 24,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#FFFFFF",
                      "fontSize": 20,
                    },
                    undefined,
                    Object {
                      "fontFamily": "simple-line-icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
          </TouchableOpacity>
        </View>
        <View
          accessibilityLabel="view--header-title-container"
          accessible={true}
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
              "marginHorizontal": 50,
            }
          }
          testID="view--header-title-container"
        >
          <Text
            accessibilityLabel="txt-test-title"
            accessible={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "#000000",
                    "fontSize": 12,
                  },
                  undefined,
                  undefined,
                  undefined,
                  Object {
                    "fontFamily": "OpenSans",
                    "fontSize": 18,
                    "lineHeight": 21,
                  },
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  false,
                  undefined,
                  undefined,
                  false,
                  false,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Object {
                    "fontWeight": "500",
                  },
                  undefined,
                  undefined,
                  undefined,
                  false,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Object {
                    "color": "#FFFFFF",
                  },
                  undefined,
                ],
              ]
            }
            testID="txt-test-title"
          >
            Test Title
          </Text>
        </View>
        <View
          accessibilityLabel="view--header-right-container"
          accessible={true}
          style={
            Object {
              "position": "absolute",
              "right": 0,
            }
          }
          testID="view--header-right-container"
        >
          <Text>
            Edit
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
